### ISSUE TITLE
Default values for extra fields of through-relations

### ISSUE THREAD

**Issue by mattii** [Vincit/objection.js#649]
First of all: thanks for this great library. It's really a pleasure to work with it! ❤️ 

Now to my question: We use join-tables to define access-levels between users and resources. This join table contains a `scope` field, which defines a user's access level (e.g. `owner`, `editor` etc.). On the resources itself, we define each group of users as separate relations (`owners`, `editors` etc.). This might seem cumbersome, but it simplifies data handling with normalized stores on the frontend a lot.

Here's a quick example, how these relations look like:
```javascript
class Project extends Model {
  static get tableName() {
    return 'projects';
  }

  static get relationMappings() {
    return {
      owners: {
        relation: Model.ManyToManyRelation,
        modelClass: `${__dirname}/user`,
        join: {
          from: 'projects.id',
          through: {
            from: 'projects_users.project_id',
            to: 'projects_users.user_id',
            extra: ['scope'],
          },
          to: 'users.id',
        },
        filter: qb => qb.where('projects_users.scope', 'owner'),
      },
      editors: {
        relation: Model.ManyToManyRelation,
        modelClass: `${__dirname}/user`,
        join: {
          from: 'projects.id',
          through: {
            from: 'projects_users.project_id',
            to: 'projects_users.user_id',
            extra: ['scope'],
          },
          to: 'users.id',
        },
        filter: qb => qb.where('projects_users.scope', 'editor'),
      },
    };
  }
}
```

Thanks to `filter`, it's really easy to define those relations and fetch the specific user-groups from the database. But when it comes to adding new users to a group like

``` javascript
await project
    .$relatedQuery('owners')
    .relate(userId)
```

it doesn't set the `scope` field to `owners`:

```sql
insert into "projects_users" ("project_id", "user_id") values (?, ?)
```

What I need, is a way to define default values for the field `scope` within the relation mapping. Do you know a way to do that?

I could imagine to extend the `extra` field API to something like

```javascript
{
  editors: {
    relation: Model.ManyToManyRelation,
    modelClass: `${__dirname}/user`,
    join: {
      from: 'projects.id',
      through: {
        from: 'projects_users.project_id',
        to: 'projects_users.user_id',
        extra: [{
            column: 'scope',
            alias: 'customPropertyName',
            onInsert: value => value || 'owner',
          },
        }],
      },
      to: 'users.id',
    },
    filter: qb => qb.where('projects_users.scope', 'editor'),
  },
}
```

Any thoughts on that? Do you think that's feasible? I would like to try a PR, if it's considered a valuable addition and there's no other way to solve it.

**Comment by koskimas** [comment]
Hi,

awesome to hear objection is working for you 👍 

There is no easy way to get this done currently. You could probably hack together something using `mergeContext`, and `$beforeInsert` but that's unreadable for anyone who isn't the original author if that hack 😄 

I'd like to make this possible. Your suggestion could work, but I'd like to avoid yet another syntax for `extra` fields. Would `onInsert` functions for `through` and root objects work for you? Something like this:

```js
{
  editors: {
    relation: Model.ManyToManyRelation,
    modelClass: `${__dirname}/user`,

    join: {
      from: 'projects.id',

      through: {
        from: 'projects_users.project_id',
        to: 'projects_users.user_id',
        extra: { customPropertyName: 'scope' },
        onInsert: (model) => (model.value = 'owner')
      },

      to: 'users.id',
    },

    filter: qb => qb.where('projects_users.scope', 'editor'),

    // You wouldn't need this for this feature, but this would solve
    // other stuff like polymorphic associations
    onInsert: (model) => (model.foo = 'bar')
  },
}
```

I don't actually know if this would be feasible because there are so many corner cases to consider (graph inserts, graph upserts etc.) but I'd like to give this a shot.

I really appreciate for the offer to work on a PR, but this would be an insane first contribution since so many corners of the code should be considered. I can start working on this if this turns out to be doable.

**Comment by mattii** [comment]
Wow, thanks for the quick reply! I totally agree with leaving `extra` as it is and I like the idea of adding it as a separate option to the `through` object. This would totally work for me.

Just let me know if I should review things. Or maybe I can at least help with documentation (BTW: another great part of this project!).

**Comment by koskimas** [comment]
Crap, we __do__ need the `through.beforeInsert` also to make `relate` operations work. Well, in your case it simply means moving the `beforeInsert` hook inside the `through` object. No need to defined it two times. I'll add a support for that next.

**Comment by mattii** [comment]
I like the consistency of `beforeInsert`! 👍 
Let me know, when I should test it. And thanks for your help and responsiveness, I highly appreciate that!

**Comment by koskimas** [comment]
I wrote some tests and it seems to work now. You can test it now using this kind of mapping:

```js
{
  editors: {
    relation: Model.ManyToManyRelation,
    modelClass: `${__dirname}/user`,

    join: {
      from: 'projects.id',

      through: {
        from: 'projects_users.project_id',
        to: 'projects_users.user_id',
        extra: { customPropertyName: 'scope' }
        beforeInsert: (model) => {
          model.customPropertyName = 'editor';
        }
      },

      to: 'users.id',
    },

    filter(qb) {
      qb.where('projects_users.scope', 'editor')
    }
  },
}
```

**Comment by mattii** [comment]
Awesome! It works! But only without a custom property name (which is totally fine in my case):

```javascript
through: {
  from: 'projects_users.project_id',
  to: 'projects_users.user_id',
  extra: ['scope'],
  beforeInsert(model) {
    model.scope = 'editor';
  },
},
```

As soon as I use `extra: { customPropertyName: 'scope' }`, it creates this SQL-command:

```sql
insert into "projects_users" ("customPropertyName", "project_id", "user_id") values (?, ?, ?)
```

**Comment by koskimas** [comment]
Oh, I thought it would work with the alias, but now that I think about it, this is how it probably should work. The alias (`customPropertyName` in this case) is used when the property is set to the parent object. I think it makes sense that the alias is not used in the "join row model instance".

**Comment by mattii** [comment]
Yes, makes sense to me. `extra` is probably rather a serialization-setting and we only related it to this new feature because of my first draft.

So from my side it's fine! I just saw that you already updated the docs. Too fast for me 😉

**Comment by koskimas** [comment]
Haha 😄 I've been meaning to add this feature for a long time but never got around to implementing it. This feature actually closes the issue number 19.

**Comment by koskimas** [comment]
Implemented in commit d8d6a8c067

**Comment by koskimas** [comment]
@mattii I just released 0.9.3 with this feature

**Comment by mattii** [comment]
Fantastic! Thank you, @koskimas!
